### PR TITLE
Clarify what character encoding is used in the returned UI strings

### PR DESCRIPTION
**Pull request by levitte**
Related to #3531

### CODE REVIEW

**Comment by levitte** [comment]
@dwmw2, @kaduk, @richsalz, you may be interested in this, please comment.

Note that this expresses the current state of affairs.

**Comment by petrovr** [comment]
Richard Levitte wrote:
> Related to #3531
>
>
>    * Clarify what character encoding is used in the returned UI strings
Hmm, it seem to me for UI encoding is irrelevant at least on Linux.
UI result cannot assume current locale as it not clear how input is 
provided - it could be from file or from wire (for instance protocol 
describes encoding not supported by C library).


Roumen

**Comment by levitte** [comment]
@petrovr, you're quite correct.  However, if you look at the changes, you'll note that I only specify the currently locale with the built in UI_METHOD `UI_OpenSSL()`, where this *is* relevant.

File: doc/man3/UI_new.pod [hunk 1]
```diff
@@ -177,11 +180,15 @@ UI_get_method() returns the UI method associated with a given UI.
...
-passphrases on the command line.
+=head1 NOTES
 
-UI_null() is a UI method that does nothing.  Its use is to avoid
-getting internal defaults for passed UI_METHOD pointers.
+The resulting strings that the built in method UI_OpenSSL() generate
+are encoded according to the current locale or (forWindows) code page.
+For applications having different demands, these strings need to be
+converted appropriately by the caller.
+For Windows, it's possible to get an UTF8 encoded string by defining
```
Comment (dwmw2) [inline doc/man3/UI_new.pod]: I'm not sure if I'd leave out the paragraph about Windows and UTF-8. Or at least try to make the caveats clearer — at first reading it seems to advocate messing with the environment variable at runtime, which has potential issues with threading. Maybe phrase it as *"if the `OPENSSL_WIN32_UTF8` environment variable is set, the built-in method `UI_OpenSSL()` will produce UTF-8 encoded strings"*?

**Comment by dwmw2** [comment]
@petrovr Sure, arbitrary files which were received from elsewhere, or data received over the wire may be in any character set under the sun. That isn't the point though.

The point here is to document what *OpenSSL* expects, especially in the cases where it uses a `UI_method` to obtain passwords. And that is the locale charset.

So... this has highlighted some other issues. If the file-decrypting functions which use a UI to obtain a password are expecting the locale charset, why do some of them look like they're expecting UTF-8? And why is there no check for `OPENSSL_WIN32_UTF8` to modify their behaviour?

This patch in itself looks good, but documenting it is of course just the start... :)

**Comment by dwmw2** [comment]
It's *almost* implicit but we should also fix the documentation of the functions which *take* a `UI_METHOD`, such the ones added in #2011, to explicitly state that they expect the locale charset.

And that helps to further deal with @petrovr 's observation. If data are coming from arbitrary files or protocols which may not be in the locale charset, then is it the responsibility of the `UI_METHOD` to fix that.

**Comment by levitte** [comment]
Guys, if would be much more helpful if the more encompassing / general / philosophical discussions were made in #3531, so we don't have to keep track of several places to follow what's actually discussed.

File: doc/man3/UI_new.pod (hunk 1 above)
Comment (levitte) [inline doc/man3/UI_new.pod]: Good point

**Comment by levitte** [comment]
Addressed all comments regarding this PR so far.

@dot-asm, care to weigh in?

**Comment by petrovr** [comment]
Richard Levitte wrote:
> @petrovr, you're quite correct.  However, if you look at the changes, you'll note that I only specify the currently locale with the built in UI_METHOD `UI_OpenSSL()`, where this *is* relevant.

"...resulting strings that the built in method UI_OpenSSL() generate are 
encoded according to the current locale or (for Windows) code page ..."

Above mean all OS-es. But I cannot see in 1.1+ code where is requirement 
for Linux. This part "..current locale.." is not correct.

I check quickly current code (perhaps I mix something) - case with read 
from terminal(tty) or  standard input (stdin). Input sequence of 
bytes(characters) is not related to "..current locale..".

Roumen

File: doc/man3/UI_new.pod [hunk 2]
```diff
@@ -183,12 +183,12 @@ UI_set_method() changes the UI method associated with a given UI.
...
 The resulting strings that the built in method UI_OpenSSL() generate
-are encoded according to the current locale or (forWindows) code page.
+are encoded according to the current locale or (for Windows) code
+page.
 For applications having different demands, these strings need to be
 converted appropriately by the caller.
-For Windows, it's possible to get an UTF8 encoded string by defining
-the environment variable C<OPENSSL_WIN32_UTF8> prior to calling
-UI_process() (one way is to define it before calling the application).
+For Windows, if the OPENSSL_WIN32_UTF8 environment variable is set,
```
Comment (dwmw2) [inline doc/man3/UI_new.pod]: Looks better; thanks.

**Comment by dwmw2** [comment]
> Above mean all OS-es. But I cannot see in 1.1+ code where is requirement for Linux. This part "..current locale.." is not correct.

No, I think this is correct. Strings read from the terminal by the default UI method will indeed be in the current locale.

Now, it is *also* the case that *all* of the rest of OpenSSL mishandles those strings, and doesn't even *know* what the current locale is. But step one is realising you have a problem. This documentation, as it stands, is correct.

**Comment by petrovr** [comment]
dwmw2 wrote:
>> Above mean all OS-es. But I cannot see in 1.1+ code where is requirement for Linux. This part "..current locale.." is not correct.
>
> No, I think this is correct. Strings read from the terminal by the default UI method will indeed be in the current locale.
>
> Now, it is *also* the case that *all* of the rest of OpenSSL mishandles those strings, and doesn't even *know* what the current locale is. But step one is realising you have a problem. This documentation, as it stands, is correct.

It is not correct because current code ui_openssl return string from tty 
or stdin as is.
There is no guarantee that encoding(CODESET) of string read from tty is 
the same as codeset of application that use ui_openssl method even on 
locale aware application.

Roumen

**Comment by levitte** [comment]
@petrovr wrote thusly:
> There is no guarantee that encoding(CODESET) of string read from tty is the same as codeset of application that use ui_openssl method even on locale aware application.

True.  However, locale settings are merely declarative, aren't they?  I.e. they declare to the applications what encoding the tty driver is assumed to send characters in...  so if we're going to check the encoding or make any conversions based on facts (or as close as facts that we can get) rather than guesses, we need to trust that the locale settings do in fact correspond to the characters we're getting in...  or have I completely misunderstood the locale workings?
(I'll admit that in this scenario, the application settings the locale on its own with `setlocale()` confuses everything)

**Comment by dwmw2** [comment]
@levitte, you are correct. The locale settings are supposed to be telling you what to expect from the terminal (and in files where otherwise unspecified, such as text files containing passwords).

In a situation where the locale setting *lies* to you, all bets are off. You don't need to worry about that case.

**Comment by petrovr** [comment]
Richard Levitte wrote:
> @petrovr wrote thusly:
>> There is no guarantee that encoding(CODESET) of string read from tty is the same as codeset of application that use ui_openssl method even on locale aware application.
>
> True.  However, locale settings are merely declarative, aren't they?  I.e. they declare to the applications what encoding the tty driver is assumed to send characters in...  so if we're going to check the encoding or make any conversions based on facts (or as close as facts that we can get) rather than guesses, we need to trust that the locale settings do in fact correspond to the characters we're getting in...  or have I completely misunderstood the locale workings?
> (I'll admit that in this scenario, the application settings the locale on its own with `setlocale()` confuses everything)

Not this case.
I mean regular call of setlocale(LC_ALL, ""). In such case 
nl_langinfo(CODESET) (if implemented) returns encoding of locale.

Issue is with command like luit that change encoding of terminal. 
Application runs in "current locale" :(

**Comment by dwmw2** [comment]
In the general case, an application can expect that the setting of `$LC_CTYPE` / `$LC_ALL` / `$LANG` correctly reflects the environment. So if `nl_langinfo()` tells you to expect UTF-8, you can expect UTF-8.

Yes, tools like luit and iconv exist which can mess around with the settings, and it's possible for them to become inconsistent and cause problems with input/output. But that's just a broken environment. You can't account for that. And you shouldn't abandon the attempt to work in a *correct* environment, just because broken environments exist.

**Comment by petrovr** [comment]
Richard Levitte wrote:
> @petrovr wrote:
>> Issue is with command like luit that change encoding of terminal.
>
> Without changing the locale accordingly?

It is by design of luit command. Application runs in "current locale".

Currently in documentation we cannot write "..locale encoding.." - code 
returns bytes read from terminal/stdin without modification.

Roumen

**Comment by levitte** [comment]
@petrovr wrote:
> Richard Levitte wrote:
> > @petrovr wrote:
> >> Issue is with command like luit that change encoding of terminal.
> >
> > Without changing the locale accordingly?
> 
> It is by design of luit command. Application runs in "current locale".
> 
> Currently in documentation we cannot write "..locale encoding.." - code 
> returns bytes read from terminal/stdin without modification.

I don't understand what you mean with "without modification".  For example, if locale for the running application ("current locale") is something like "sv_SE.utf8", it's reasonable to receive "räksmörgås" as  0x72 0xc3 0xa4 0x6b 0x73 0x6d 0xc3 0xb6 0x72 0x67 0xc3 0xa5 0x73, but if the locale is set to "sv_SE.iso88591", the same string can reasonably be expected to be 0x72 0xe4 0x6b 0x73 0x6d 0xf6 0x72 0x67 0xe5 0x73.  Makes sense?

**Comment by petrovr** [comment]
dwmw2 wrote:
>
> A string *requires* an encoding.
No until is you would like to show to humans.

>  A byte sequence without an encoding *cannot* be a string.

Unfortunately it is.


> I hereby define "locale encoding" to mean the encoding which is used for strings read from or written to the terminal, and from the command line and most otherwise unlabelled text files in the file system.

Yes. The same here. It is expected textual information to use encoding 
specified by locale.

> In a sane setup where luit and other things haven't been misused, we will expect `nl_langinfo(CODESET)` to correctly tell us what the locale encoding is.

Sane setup between various OS-es? Localization (locales) are not 
standardized.
Locale is more then encoding for text. "Locale encoding" is too 
restrictive. OS may not support some languages or regions.  Encoding is 
language and territory neutral.


> Now... the string that UI_OpenSSL() reads from the terminal and gives you *ABSOLUTELY IS* in the locale encoding.

No. It is not true!

Where is the code .
The current code returns read bytes as is (!WIN32, !DOS).

You would like to describe non-existing functionality  like : 
C-function fgets  returns characters in locale encoding!!!


> Can we be done with this line of argument now please?
>

Unfortunately fgets returns bytes.


Roumen

**Comment by levitte** [comment]
Ah, I think I understand what @petrovr is talking about, and this fixup should be satisfactory, I hope:

``` diff
diff --git a/doc/man3/UI_new.pod b/doc/man3/UI_new.pod
index 436030af2a..5b98cf8d0d 100644
--- a/doc/man3/UI_new.pod
+++ b/doc/man3/UI_new.pod
@@ -183,12 +183,13 @@ UI_set_method() changes the UI method associated with a given UI.
 =head1 NOTES
 
 The resulting strings that the built in method UI_OpenSSL() generate
-are encoded according to the current locale or (for Windows) code
-page.
+are assumed to be encoded according to the current locale or (for
+Windows) code page.
 For applications having different demands, these strings need to be
 converted appropriately by the caller.
 For Windows, if the OPENSSL_WIN32_UTF8 environment variable is set,
-the built-in method UI_OpenSSL() will produce UTF-8 encoded strings.
+the built-in method UI_OpenSSL() will produce UTF-8 encoded strings
+instead.
 
 =head1 COPYRIGHT
 
```

**Comment by petrovr** [comment]
Richard Levitte wrote:
> [SNIP]
>
> I don't understand what you mean with "without modification".  For example, if locale for the running application ("current locale") is something like "sv_SE.utf8",
 > it's reasonable to receive "räksmörgås" as  0x72 0xc3 0xa4 0x6b 0x73 
0x6d 0xc3 0xb6 0x72 0x67 0xc3 0xa5 0x73, >

If is stored 0x72 0xc3 0xa4 0x6b 0x73 0x6d 0xc3 0xb6 0x72 0x67 0xc3 0xa5 
0x73 you will receive that same sequence even if you locale is 
sv_SE.iso88591.

 > but if the locale is set to "sv_SE.iso88591", the same string can 
reasonably be expected to be 0x72 0xe4 0x6b 0x73 0x6d 0xf6 0x72 0x67 
0xe5 0x73.  Makes sense?
>

"without modification" mean that sequence "0x72 0xe4 0x6b 0x73 0x6d 0xf6 
0x72 0x67 0xe5 0x73" will be received from tty/stdin whatever encoding 
is set for locale on reader even on locale aware applications.


With "locale encoding" is expected read from stream with locale_t as 
argument.

**Comment by levitte** [comment]
> "without modification" mean that sequence "0x72 0xe4 0x6b 0x73 0x6d 0xf6 
> 0x72 0x67 0xe5 0x73" will be received from tty/stdin whatever encoding 
> is set for locale on reader even on locale aware applications.

This is not true.  I tried the following program just now:

``` C
#include <stdio.h>
#include <openssl/ui.h>

void hexdump(FILE *out, char *str)
{
  while (*str != '\0') {
    fprintf(out, " 0x%02x", *(unsigned char *)str++);
  }
  fprintf(out, "\n");
}

int main(int argc, char *argv[])
{
  UI *ui = UI_new();
  char *prompt = NULL;
  char pass[1024];

  if (ui == NULL) {
    exit(1);
  }

  if ((prompt = UI_construct_prompt(ui, "pass phrase", "test")) != NULL
      && UI_add_input_string(ui, prompt, UI_INPUT_FLAG_DEFAULT_PWD,
			     pass, 0, sizeof(pass) - 1)
      && UI_process(ui) >= 0) {
    hexdump(stdout, pass);
  }

  OPENSSL_free(prompt);
  UI_free(ui);
  exit(0);
}
```

Here's the result when I type in "räksmörgås"

```
: ; cc foo.c -lcrypto
: ; ./a.out 
Enter pass phrase for test:
 0x72 0xc3 0xa4 0x6b 0x73 0x6d 0xc3 0xb6 0x72 0x67 0xc3 0xa5 0x73
: ; locale
LANG=sv_SE.utf8
LANGUAGE=
LC_CTYPE="sv_SE.utf8"
LC_NUMERIC="sv_SE.utf8"
LC_TIME="sv_SE.utf8"
LC_COLLATE="sv_SE.utf8"
LC_MONETARY="sv_SE.utf8"
LC_MESSAGES="sv_SE.utf8"
LC_PAPER="sv_SE.utf8"
LC_NAME="sv_SE.utf8"
LC_ADDRESS="sv_SE.utf8"
LC_TELEPHONE="sv_SE.utf8"
LC_MEASUREMENT="sv_SE.utf8"
LC_IDENTIFICATION="sv_SE.utf8"
LC_ALL=
```

The reason for this is that I ran it in a urxvt window, which does generate UTF8 encoded strings, and of course, my locale is set accordingly.

If you get the string 0x72 0xe4 0x6b 0x73 0x6d 0xf6 0x72 0x67 0xe5 0x73, then the correct locale would be "sv_SE.iso88591" (or really, anything that declares ISO 8859-1).  Anything else is *incorrect*

**Comment by dwmw2** [comment]
Trying desperately to make sense of what you're saying... Let's try another approach.

You receive the bytes `0x72 0xe4 0x6b 0x73 0x6d 0xf6 0x72 0x67 0xe5 0x73` from the UI method which read those bytes from the console. Those bytes represent a string of characters... How should you correctly convert those characters to BMPString for the purpose of creating a PKCS#12 file?

**Comment by petrovr** [comment]
Richard Levitte wrote:
>> "without modification" mean that sequence "0x72 0xe4 0x6b 0x73 0x6d 0xf6
>> 0x72 0x67 0xe5 0x73" will be received from tty/stdin whatever encoding
>> is set for locale on reader even on locale aware applications.
>
> This is not true.  I tried the following program just now:

I compile you code into binary test-ui-method

$ LANG=bg_BG.UTF-8 locale charmap
UTF-8
$ LANG=bg_BG.KOI8-R locale charmap
KOI8-R
$ locale
LANG=bg_BG.UTF-8
...

So terminal encoding is UTF-8.

On command line I type my name in my language and run you program in 
"current locale" (read standard input)
$ ( echo румен | LANG=bg_BG.KOI8-R iconv -f UTF-8 ) | setsid 
./test-ui-method
Enter pass phrase for test:
  0xd2 0xd5 0xcd 0xc5 0xce


$ echo румен | hexdump -C
00000000  d1 80 d1 83 d0 bc d0 b5  d0 bd 0a                 |...........|
0000000b

If UI method reads password in "current locale" you test case should 
produce 0xd1 0x80 ....

[SNIP]

That way I cannot understand text that say result is "current local 
encoding". It is too strict but above tests shows "read as is" = what 
fgets read from stream this is returned value for UI without new line.

I have no idea how to write facts to be readable by users.

**Comment by petrovr** [comment]
dwmw2 wrote:
> Trying desperately to make sense of what you're saying... Let's try another approach.
>
> You receive the bytes `0x72 0xe4 0x6b 0x73 0x6d 0xf6 0x72 0x67 0xe5 0x73` from the UI method which read those bytes from the console. Those bytes represent a string of characters... How should you correctly convert those characters to BMPString for the purpose of creating a PKCS#12 file?

Out of scope .
#3531

**Comment by petrovr** [comment]
Richard Levitte wrote:
> Ah, I think I understand what @petrovr is talking about, and this fixup should be satisfactory, I hope:
>
[SNIP]

Perhaps "...assumed to be encoded according to the current locale..." is 
good compromise for users.

As above sentence is not strict current implementation of UI method or 
use of UI method could remain unmodified.

Roumen

**Comment by dwmw2** [comment]
```
$ ( echo румен | LANG=bg_BG.KOI8-R iconv -f UTF-8 ) | setsid
```
This is just broken nonsense. You are deliberately converting charset here, from UTF-8 to KOI8-R. So the string "румен" ends up rendered as the bytes `0xd2 0xd5 0xcd 0xc5 0xce`. You are feeding those bytes as stdin to a program running in the unmodified default UTF-8 locale, which *obviously* doesn't know how to make sense of them.

I can pass input through `iconv -t EBCDIC` too, and it doesn't actually prove a useful point...

**Comment by levitte** [comment]
@petrovr wrote:
> On command line I type my name in my language and run you program in 
> "current locale" (read standard input)
> $ ( echo румен | LANG=bg_BG.KOI8-R iconv -f UTF-8 ) | setsid ./test-ui-method
> Enter pass phrase for test:
>   0xd2 0xd5 0xcd 0xc5 0xce

Yes, of course, since you're converting it to the KOI8-R codeset, that's what you'll get.  *BUT*, you're lying about the codeset to `setsid`.  Since locale setting is a *declaration*, the correct thing to do is this:

```
$ ( echo румен | LANG=bg_BG.KOI8-R iconv -f UTF-8 ) | ( LANG=bg_BG.KOI8-R setsid ./a.out )
```

> If UI method reads password in "current locale" you test case should 
> produce 0xd1 0x80 ....

No, it should not.  You should declare your locale correctly, corresponding to what your tty driver is expected to feed to you.

**Comment by petrovr** [comment]
My point was to show how openssl UI method works.
Command  ....  | ( LANG=NotUsed setsid ./a.out ) produce the same result.

**Comment by dwmw2** [comment]
The UI method just gives out the same bytes it receives. If those are not in the charset specified by `$LANG` that's because the environment (the extremely contrived one, in this case) is broken. This whole line of debate is silly.

**Comment by dwmw2** [comment]
OMG I tried this and it didn't give me the bytes I expected either!
```
 echo румен | rev | ./a.out
```

**Comment by levitte** [comment]
> My point was to show how openssl UI method works.
> Command  ....  | ( LANG=NotUsed setsid ./a.out ) produce the same result.

Yes, correct.  Again, the locale *only declares what programs should expect*.  If you're setting a locale that doesn't match the bytes coming from the TTY, *your locale settings are wrong*

How many times do we need to say this?

**Review by richsalz — APPROVED** [review]
for master and any backports you want to do.

**Comment by petrovr** [comment]
Richard Levitte wrote:
>> My point was to show how openssl UI method works.
>> Command  ....  | ( LANG=NotUsed setsid ./a.out ) produce the same result.
>
> Yes, correct.  Again, the locale *only declares what programs should expect*.  If you're setting a locale that doesn't match the bytes coming from the TTY, *your locale settings are wrong*
>
> How many times do we need to say this?

Quote from manual :   fgetws (but, sizeof(buf), stdin);
NOTES
        The behavior of fgetws() depends on the LC_CTYPE category of the 
current locale.


fgetws is function that depend from "current locale" while fgets is 
locale neutral.

The difference is visible be example:
----
#include <wchar.h>
#include <locale.h>
#include <stdlib.h>
#include <stdio.h>

int
main() {
   wchar_t buf[100];

   setlocale(LC_ALL, "");

   fgetws(buf, sizeof(buf), stdin);
   wprintf(L"%ls\n", buf);

   return 0;
}
----

And the test command that shows the we MUST set "locale that match the 
bytes coming from the TTY" :
(echo румен | LANG=bg_BG iconv -f UTF-8) | (LANG=bg_BG 
./test-locale-input | iconv -f CP1251)

Sorry that I have to repeat that current UI method is locale neutral .

Roumen

**Comment by levitte** [comment]
Merged.

master: 789d6dddecbc231f5c858d203aab318cf81676fd
1.1.0: b024891f7f84d657d1c9b1f0d364dc1c3c42f411

**Comment by dwmw2** [comment]
Roumen, what term would you use to describe "the charset encoding of stdin/stdout"?

fgetws() is just converting from that charset (which the result of call the locale charset) implicitly.

**Comment by levitte** [comment]
> fgetws() is just converting from that charset (which the rest of us call the locale charset) implicitly.

Hmmm?  From the man page, it looks to me like it expects to receive an exact number of wide chars on the line, period.  Do I misunderstand?

**Comment by dwmw2** [comment]
Wide chars are UCS2. So fgetws() implicitly does the conversion of the raw input bytes from the filedescriptor, from $LC_CTYPE to UCS2. Basically the same conversion we need to do to get UCS2.

Not quite sure why he was bringing it up; just another red herring. We aren't using fgetws().

**Comment by dwmw2** [comment]
@petrovr it really is starting to look like you're just trolling, refusing to follow basic questions that would lead to enlightenment, then throwing in other random red herrings. But just in case, I'll try one more time...
```
 fgetws(buf, sizeof(buf), stdin);
 wprintf(L"%ls\n", buf);
```

Run in an 8-bit locale, this program is just a slightly CPU-hogging version of `cat`.

You're feeding it the five bytes `0xf0 0xf3 0xec 0xe5 0xed` which in the CP1251 encoding represents "румен".

*(Why *are* you using CP1251 here when the example was KOI8-R before, btw? That also just makes it look like needlessly sowing confusion...)*

We observed earlier that `fgets()` simply reads those five bytes directly from stdin and one needs to know the encoding in order to interpret them — do they mean `"ПСЛЕМ"` (KOI8-R), or `"ðóìåí"` (ISO8859-1) or `"đķėåí"` (ISO8859-4), etc...

But you're not using `fgets()` so you don't see those 5 bytes. You're now using `fgetws()`. As documented, that implicitly performs the conversion from the legacy character set, to wide characters which are UCS-2. And then you immediately use `wprintf()` which converts them right back to the same legacy character set to send them to stdout. So if we augment your test to look like this:

```
	fgetws(buf, sizeof(buf), stdin);
	fprintf(stderr, "In %s:", nl_langinfo(CODESET));
	for (i=0; buf[i]; i++)
		fprintf(stderr, " U+%x", buf[i]);
	fprintf(stderr, "\n");
	wprintf(L"%ls\n", buf);
```
We get...
```
 $ echo -n 'румен' | iconv -t CP1251 > fivebytes
 $ hexdump -C fivebytes
00000000  f0 f3 ec e5 ed                                    |.....|
 $ LC_ALL=bg_BG ./a.out < fivebytes | hexdump -C
In CP1251: U+440 U+443 U+43c U+435 U+43d
00000000  f0 f3 ec e5 ed 0a                                 |......|
$ LC_ALL=en_GB ./a.out < fivebytes | hexdump -C
In ISO-8859-1: U+f0 U+f3 U+ec U+e5 U+ed
00000000  f0 f3 ec e5 ed 0a                                 |......|
$ LC_ALL=ru_RU ./a.out < fivebytes | hexdump -C
In ISO-8859-5: U+2116 U+453 U+44c U+445 U+44d
00000000  f0 f3 ec e5 ed 0a                                 |......|
 $ LC_ALL=ru_RU.KOI8-R ./a.out < fivebytes | hexdump -C
In KOI8-R: U+41f U+421 U+41b U+415 U+41c
00000000  f0 f3 ec e5 ed 0a                                 |......|
```

You'll note they all interpret them as different characters, and (to get back to the point slightly) the BMPString used for generating the PKCS#12 key would be different in each case for the same five bytes of input. But if you perform a conversion from $LC_CTYPE to UCS2 and back again... what a surprise, all of them output the same five bytes on stdout! And the \n that you added, of course.

(If you run a.out in a UTF-8 locale, things are different because those five bytes aren't actually valid UTF-8 so they *can't* be interpreted.)

Your example here isn't actually demonstrating anything that we didn't already know.

If you read from stdin with "normal" char functions as `UI_OpenSSL()` does,, you get raw bytes and you need to know the charset they are in — which we shall call "the locale charset" because all else is just messed up beyond belief.

If you use `fgetws()` then the C library automatically does that conversion for you and gives you the result as UCS-2 `wchar_t`, relieving you of the need to process it for yourself.

None of this makes any sense as an objection to the patch in this PR, which is thankfully already merged. If you're saying *"hey guys, how about we think about using `fgetws()` in future and using UCS-2 and `wchar_t` internally instead of `char`"* then that would at least be *coherent* although I'd probably argue against it. But I'm not even sure that is what you're saying; I'm just clutching at straws to find some way that you *aren't* just trolling...

**Pull request closed without merging**